### PR TITLE
make start, tend optional in get_fullycoherent_twoF()

### LINKED ISSUE #167: TransientGridSearch problem with minStartTime=None, maxStartTime=None
`TransientGridSearch` has a problem with `minStartTime=None`, `maxStartTime=None`, trying to initialise from t0 from `int(None)`. Even with those options, a proper start time should be available, read from SFTs, through `ComputeFstat.__init__()`, but that is called too late and the result not propagated back to the `self` itself.

My first idea was to move the `inititate_search_object()` call from `run()` to `__init__()` but that's a bad idea because it doesn't play well with reusing results from disk. Will instead need some slightly smarter if-first-run-or-not if/else logic.

### PR DESCRIPTION
**Pull request by dbkeitel**
This is another case where fixing something small ballooned into reworking a core feature, because working around it was just too messy.

The main change is to make `tstart` and `tend` optional arguments in the core function `get_fullycoherent_twoF()`, falling back to `self.minStartTime` and `self.maxStartTime` if they're `None`, which is sufficient for the basic CW classes `GridSearch` and `MCMCSearch` and only needs to be overridden point-by-point for some of the transient and glitch classes.

- requires lots of tracking changes in grid and mcmc searches
- change grid searches to use named-column ndarrays internally to avoid fragile numeric indexing
- partially clean up keys storage in MCMC searches
- change helper_functions.get_doppler_params_output_format() to return dict
- amend tests and examples
~- WIP: SlidingWindow and FrequencySlidingWindow broken for now~
- closes #167

### CODE REVIEW

File: examples/grid_examples/PyFstat_example_grid_search_F0.py [hunk 1]
```diff
@@ -69,11 +69,10 @@
...
 search.run()
 
+print("Plotting 2F(F0)...")
 fig, ax = plt.subplots()
-xidx = search.keys.index("F0")
-frequencies = np.unique(search.data[:, xidx])
-twoF = search.data[:, -1]
-
+frequencies = search.data["F0"]
+twoF = search.data["twoF"]
```
Comment (dbkeitel) [inline examples/grid_examples/PyFstat_example_grid_search_F0.py:75]: this is the basic example for why the change to the internal data structures makes usage easier

File: pyfstat/core.py [hunk 2]
```diff
@@ -851,8 +851,6 @@ def estimate_min_max_CoverFreq(self):
 
     def get_fullycoherent_twoF(
         self,
-        tstart,
-        tend,
         F0,
```
Comment (dbkeitel) [inline pyfstat/core.py:854]: It might be a good further step to change this function to **all keyword args only**, so that there can be no confusion about argument order.

File: pyfstat/core.py [hunk 3]
```diff
@@ -863,8 +861,11 @@ def get_fullycoherent_twoF(
         ecc=None,
         tp=None,
         argp=None,
+        tstart=None,
+        tend=None,
     ):
         """ Returns twoF or ln(BSGL) fully-coherently at a single point """
+        # print("[F0,F1,F2,Alpha,Delta,tstart,tend] = ", [F0,F1,F2,Alpha,Delta,tstart,tend])
```
Comment (dbkeitel) [inline pyfstat/core.py:868]: oops, surviving debug statement, will kill on next push

File: pyfstat/grid_based_searches.py [hunk 4]
```diff
@@ -101,14 +101,13 @@ def __init__(
             os.mkdir(outdir)
         self.set_out_file()
         self.search_keys = ["F0", "F1", "F2", "Alpha", "Delta"]
-        self.input_keys = ["minStartTime", "maxStartTime"] + self.search_keys
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:104]: this one seemed superfluous

File: pyfstat/grid_based_searches.py [hunk 5]
```diff
@@ -101,14 +101,13 @@ def __init__(
...
-        self.keys = self.input_keys.copy()
+        self.output_keys = self.search_keys.copy()
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:104]: And `keys` seemed too ambiguous, what it actually already was was the list for the output `data` array, so I renamed it appropriately.

File: pyfstat/grid_based_searches.py [hunk 6]
```diff
@@ -119,7 +118,7 @@ def _get_search_ranges(self):
         else:
             return None
 
-    def inititate_search_object(self):
+    def _initiate_search_object(self):
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:121]: I don't think this should ever be called directly by an outside user.

File: pyfstat/grid_based_searches.py [hunk 7]
```diff
@@ -157,11 +156,11 @@ def inititate_search_object(self):
...
-
-            def cut_out_tstart_tend(*vals):
-                return self.search.get_semicoherent_det_stat(*vals[2:])
-
-            self.search.get_det_stat = cut_out_tstart_tend
+            self.search.get_det_stat = self.search.get_semicoherent_det_stat
+        # make sure to overwrite the min/max starttime in case the user
+        # passed None and they were read from SFTs
+        self.minStartTime = self.search.minStartTime
+        self.maxStartTime = self.search.maxStartTime
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:163]: This is what should solve #167 and was impossible to do without the bigger changes or a similarly huge reordering.

File: pyfstat/grid_based_searches.py [hunk 8]
```diff
@@ -157,11 +156,11 @@ def inititate_search_object(self):
                 detectors=self.detectors,
                 injectSources=self.injectSources,
             )
-
-            def cut_out_tstart_tend(*vals):
-                return self.search.get_semicoherent_det_stat(*vals[2:])
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:162]: ugly and no longer needed :)

File: pyfstat/grid_based_searches.py [hunk 9]
```diff
@@ -229,29 +234,29 @@ def check_old_data_is_okay_to_use(self):
             )
 
         logging.info("Loading old data from '{:s}'.".format(self.out_file))
-        old_data = np.atleast_2d(np.genfromtxt(self.out_file, delimiter=" "))
-        # need to convert any "None" entries in input_data array safely to 0s
-        # to make np.allclose() work reliably
-        new_data = np.nan_to_num(self.input_data.astype(np.float64))
+        old_data = helper_functions.read_txt_file_with_header(self.out_file, names=True)
+        # old_data = np.atleast_2d(old_data) # FIXME do we still need this?
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:238]: TODO: check a single-point test

File: pyfstat/grid_based_searches.py [hunk 10]
```diff
@@ -271,48 +276,51 @@ def run(self, return_data=False):
...
+        )
+        data = np.zeros(len(self.input_data), dtype=output_dtype)
+        for n, vals in enumerate(
+            tqdm(iterable, total=getattr(self, "total_iterations", None))
+        ):
             detstat = self.search.get_det_stat(*vals)
             thisCand = list(vals) + [detstat]
-            data.append(thisCand)
+            for k, key in enumerate(self.output_keys):
+                data[key][n] = thisCand[k]
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:294]: I didn't find a nice one-liner for this due to the general clunkiness of numpy structured arrays. Maybe we want to check timing on this.

File: pyfstat/grid_based_searches.py [hunk 11]
```diff
@@ -774,22 +778,32 @@ def run(self, return_data=False):
...
 
-        data = np.array(data, dtype=np.float)
         if return_data:
             return data
         else:
-            self.save_array_to_disk(data)
             self.data = data
+            self.save_array_to_disk()
+
+    def get_transient_fstat_map_filename(self, param_point):
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:787]: This was supposed to be its own commit, but probably too bothersome to extract now that I've somehow squashed it in... should amend commit message to acknowledge the addition, though.

File: pyfstat/grid_based_searches.py [hunk 12]
```diff
@@ -1132,17 +1145,15 @@ def __init__(
...
-        fmt += ["%.9g"]  # for detection statistic
+        fmt = helper_functions.get_doppler_params_output_format(self.output_keys)
+        fmt["delta_F0"] = "%.16g"
+        fmt["delta_F1"] = "%.16g"
+        fmt["tglitch"] = "%d"
+        fmt[self.detstat] = "%.9g"
         return fmt
 
 
+# FIXME below here not fixed for [minStartTime,maxStartTime] removal yet
```
Comment (dbkeitel) [inline pyfstat/grid_based_searches.py:1156]: TODO I'm not yet sure enough what those classes are supposed to be doing to fix them...

File: pyfstat/mcmc_based_searches.py [hunk 13]
```diff
@@ -283,28 +283,17 @@ def logp(self, theta_vals, theta_prior, theta_keys, search):
         return np.sum(H)
 
     def logl(self, theta, search):
+        in_theta = copy.copy(self.fixed_theta)
         for j, theta_i in enumerate(self.theta_idxs):
-            self.fixed_theta[theta_i] = theta[j]
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:287]: It seemed really ugly to overwrite `self.fixed_theta` in each iteration, so I cleaned this up along the way, here and for several child classes.

File: pyfstat/mcmc_based_searches.py [hunk 14]
```diff
@@ -283,28 +283,17 @@ def logp(self, theta_vals, theta_prior, theta_keys, search):
...
+        twoF = search.get_fullycoherent_twoF(*in_theta)
         return twoF / 2.0 + self.likelihoodcoef
 
     def _unpack_input_theta(self):
-        full_theta_keys = ["F0", "F1", "F2", "Alpha", "Delta"]
-        if self.binary:
-            full_theta_keys += ["asini", "period", "ecc", "tp", "argp"]
-        full_theta_keys_copy = copy.copy(full_theta_keys)
-
-        full_theta_symbols = [
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:299]: also removed redundant definitions of these

File: pyfstat/mcmc_based_searches.py [hunk 15]
```diff
@@ -283,28 +283,17 @@ def logp(self, theta_vals, theta_prior, theta_keys, search):
...
-            r"$f$",
-            r"$\dot{f}$",
-            r"$\ddot{f}$",
-            r"$\alpha$",
-            r"$\delta$",
-        ]
+        self.full_theta_keys = ["F0", "F1", "F2", "Alpha", "Delta"]
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:293]: now storing these in `self` so they can be indexed over

File: pyfstat/mcmc_based_searches.py [hunk 16]
```diff
@@ -3288,40 +3284,25 @@ def _initiate_search_object(self):
...
+        for j, theta_i in enumerate(self.theta_idxs):
+            in_theta[theta_i] = theta[j]
+        t0idx = self.full_theta_keys.index("transient_tstart")
+        t1idx = self.full_theta_keys.index("transient_duration")
+        in_theta[t1idx] = in_theta[t0idx] + in_theta[t1idx]  # endtime = t0+tau
+        if in_theta[t1idx] > self.maxStartTime:
             return -np.inf
-        twoF = search.get_fullycoherent_twoF(*in_theta)
+        twoF = search.get_fullycoherent_twoF(
+            *in_theta[:-2], tstart=in_theta[t0idx], tend=in_theta[t1idx]
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:3296]: TODO: one surviving instance of hard-coded numeric indices, I should fix that

**Comment by dbkeitel** [comment]
Solved rebase conflicts against current master. So far it still fails two tests:
```
FAILED tests.py::TestComputeFstat::test_cumulative_twoF - RuntimeError: Internal function call failed: Invalid argument
FAILED tests.py::TestGridSearch::test_sliding_window - AttributeError: 'FrequencySlidingWindow' object has no attribute 'output_keys'
```
and probably some examples.

Code needs a thorough self-review too.

File: pyfstat/mcmc_based_searches.py [hunk 17]
```diff
@@ -3322,40 +3318,34 @@ def _initiate_search_object(self):
...
+        )
+        # FIXME: this can be simplified when changing all theta lists to dicts
+        t0idx = self.theta_keys.index("transient_tstart")
+        t1idx = self.theta_keys.index("transient_duration")
+        in_theta["tstart"] = theta[t0idx]
+        in_theta["tend"] = theta[t0idx] + theta[t1idx]  # endtime = t0+tau
+        if in_theta["tend"] > self.maxStartTime:
             return -np.inf
-        twoF = search.get_fullycoherent_twoF(*in_theta)
+        twoF = search.get_fullycoherent_twoF(**in_theta)
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:3340]: Ideally this is how I think the function should always be called, and a follow-up PR could change the whole MCMC module to always use dicts and no lists for all its `theta` needs and then do it like here.

File: pyfstat/grid_based_searches.py (hunk 9 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:247]: proper solution: fix `helper_functions.read_txt_file_with_header`

**Comment by dbkeitel** [comment]
After rebasing on #215 which removes `FrequencySlidingWindow` all tests pass.
However, [running examples](https://github.com/PyFstat/PyFstat/runs/1333945558?check_suite_focus=true) gives two failures:
```
Failed to run 2/23 example scripts: ['/home/runner/work/PyFstat/PyFstat/examples/transient_examples/PyFstat_example_long_transient_MCMC_search.py', '/home/runner/work/PyFstat/PyFstat/examples/mcmc_examples/PyFstat_example_fully_coherent_MCMC_search.py.bk']
```

**Comment by dbkeitel** [comment]
* `PyFstat_example_fully_coherent_MCMC_search.py.bk` was obviously a spurious file that shouldn't have been in the commit
* `PyFstat_example_long_transient_MCMC_search.py` failed because it searches over tau only (fixed t0), a situation which was not covered in the tests. I've now fixed this and extended the tests.

**Comment by dbkeitel** [comment]
@Rodrigo-Tenorio Now here's the big monster to review. I've already left a few comments on crucial parts of the diff, but there's a lot more to it...

**Review by Rodrigo-Tenorio — COMMENTED** [review]
Summary of this changes: Make `tstart` and `tend` optional in `get_fullycoherent_fstat`. This make sense, since these options are not what one would think they are, since a window needs to be specified by the time this is called.

File: pyfstat/helper_functions.py [hunk 18]
```diff
@@ -478,23 +477,21 @@ def get_doppler_params_output_format(keys):
         "tp",
         "argp",
     ]
-
-    for k in keys:
-        if k in doppler_keys:
-            fmt += [CFSv2_fmt]
+    fmt = {k: CFSv2_fmt for k in keys if k in doppler_keys}
```
Comment (Rodrigo-Tenorio) [inline pyfstat/helper_functions.py:480]: Reminder to check where is this function used

File: pyfstat/mcmc_based_searches.py (hunk 13 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/mcmc_based_searches.py:288]: Is fixed_theta a dictionary? Why not just `self.fixed_theta.copy()`?

File: pyfstat/mcmc_based_searches.py [hunk 19]
```diff
@@ -284,28 +284,17 @@ def logp(self, theta_vals, theta_prior, theta_keys, search):
...
 
     def logl(self, theta, search):
+        in_theta = copy.copy(self.fixed_theta)
         for j, theta_i in enumerate(self.theta_idxs):
-            self.fixed_theta[theta_i] = theta[j]
-        twoF = search.get_fullycoherent_twoF(
-            self.minStartTime, self.maxStartTime, *self.fixed_theta
-        )
+            in_theta[theta_i] = theta[j]
+        twoF = search.get_fullycoherent_twoF(*in_theta)
```
Comment (Rodrigo-Tenorio) [inline pyfstat/mcmc_based_searches.py:290]: I feel quite uncomfortable having this call without keywords; but you said that would be a future rework, so fine by now.

File: pyfstat/mcmc_based_searches.py [hunk 20]
```diff
@@ -1731,15 +1720,15 @@ def _check_old_data_is_okay_to_use(self):
...
+        fmt["twoF"] = "%.9g"
         return fmt
 
     def export_samples_to_disk(self):
         self.samples_file = os.path.join(self.outdir, self.label + "_samples.dat")
         logging.info("Exporting samples to {}".format(self.samples_file))
         header = "\n".join(self.output_file_header)
         header += "\n" + " ".join(self.theta_keys) + " twoF"
-        outfmt = self.get_savetxt_fmt()
+        outfmt = list(self.get_savetxt_fmt().values())
```
Comment (Rodrigo-Tenorio) [inline pyfstat/mcmc_based_searches.py:1731]: Ok!

File: pyfstat/mcmc_based_searches.py (hunk 17 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/mcmc_based_searches.py:3347]: Agree

File: pyfstat/mcmc_based_searches.py [hunk 21]
```diff
@@ -3322,40 +3318,41 @@ def _initiate_search_object(self):
...
-        full_theta_keys = [
-            "transient_tstart",
-            "transient_duration",
-            "F0",
-            "F1",
-            "F2",
-            "Alpha",
-            "Delta",
-        ]
+        self.full_theta_keys = ["F0", "F1", "F2", "Alpha", "Delta"]
```
Comment (Rodrigo-Tenorio) [inline pyfstat/mcmc_based_searches.py:3351]: I've seen this same list hard coded too many times...

**Review by Rodrigo-Tenorio — COMMENTED** [review]
Excluding `grid` and `tests`, changes look fine to me

File: pyfstat/mcmc_based_searches.py (hunk 13 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py:288]: It should be a dict, but as elsewhere: future work.

File: pyfstat/helper_functions.py (hunk 18 above)
Comment (dbkeitel) [reply pyfstat/helper_functions.py:480]:
```
grep -n get_doppler_params_output_format pyfstat/*.py
pyfstat/grid_based_searches.py:340:        fmt = helper_functions.get_doppler_params_output_format(self.output_keys)
pyfstat/grid_based_searches.py:845:        fmt = helper_functions.get_doppler_params_output_format(self.output_keys)
pyfstat/grid_based_searches.py:968:        fmt = helper_functions.get_doppler_params_output_format(self.output_keys)
pyfstat/helper_functions.py:464:def get_doppler_params_output_format(keys):
pyfstat/mcmc_based_searches.py:1722:        fmt = helper_functions.get_doppler_params_output_format(self.theta_keys)
pyfstat/mcmc_based_searches.py:2485:        fmt = helper_functions.get_doppler_params_output_format(
pyfstat/mcmc_based_searches.py:2488:        fmt = helper_functions.get_doppler_params_output_format(self.theta_keys)
pyfstat/mcmc_based_searches.py:3388:        fmt = helper_functions.get_doppler_params_output_format(self.theta_keys)
```

All use it in `dict` fashion, but `pyfstat/mcmc_based_searches.py:2485` + `pyfstat/mcmc_based_searches.py:2488` makes no sense, let me see what and why I did there.

File: pyfstat/mcmc_based_searches.py [hunk 22]
```diff
@@ -2494,10 +2485,14 @@ def get_savetxt_fmt(self):
         fmt = helper_functions.get_doppler_params_output_format(
             [k.split("_")[1] for k in self.theta_keys if k.startswith("delta_F")]
         )
+        fmt = helper_functions.get_doppler_params_output_format(self.theta_keys)
```
Comment (dbkeitel) [inline pyfstat/mcmc_based_searches.py:2488]: Why am I calling this twice...? :thinking:

File: pyfstat/mcmc_based_searches.py (hunk 19 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py:290]: Yeah, at least it's no worse than before.

File: pyfstat/mcmc_based_searches.py (hunk 21 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py:3351]: Maybe it should be an attribute of the `ComputeFStatistic` core class, `default_doppler_parameter_keys` or something...? But that should be another separate change, I'd say.

File: pyfstat/mcmc_based_searches.py (hunk 20 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py:1731]: Actually, not ok! This will result in `outfmt` in the order the dict was filled, which does not need to be the same as the order of samples columns. This currently causes a bug at least for `MCMCGlitchSearch`.

File: pyfstat/mcmc_based_searches.py (hunk 22 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py:2501]: ...so that I would have reason to fix a hidden output bug. ;)

File: pyfstat/helper_functions.py (hunk 18 above)
Comment (dbkeitel) [reply pyfstat/helper_functions.py:480]: Duplication resolved, but it was helpful in noticing and fixing a sorting issue with the final output formats list.

File: pyfstat/mcmc_based_searches.py (hunk 20 above)
Comment (dbkeitel) [reply pyfstat/mcmc_based_searches.py]: I think the `MCMCGlitchSearch` glitch output was already buggy on master, see #219, though this PR initially made it worse. ;). Should now be fixed in 662d47217d97ef528663f121f0e02f432b0e3fad but I still need to double-check if similar issues appear in the grid classes.

File: pyfstat/core.py (hunk 2 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/core.py:864]: Agree. Future work.

File: pyfstat/mcmc_based_searches.py (hunk 20 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/mcmc_based_searches.py]: This looks like on of those "Here's yet another way of writing data down!" which #191 should take care of. I'll mark this.

File: pyfstat/mcmc_based_searches.py [hunk 23]
```diff
@@ -3378,21 +3388,23 @@ def _unpack_input_theta(self):
...
         self.theta_idxs = [self.theta_idxs[i] for i in idxs]
         self.theta_symbols = [self.theta_symbols[i] for i in idxs]
         self.theta_keys = [self.theta_keys[i] for i in idxs]
 
-    def get_savetxt_fmt(self):
-        fmt = []
+        self.output_keys = self.theta_keys.copy()
+        self.output_keys += ["logBSGL" if self.BSGL else "twoF"]
+
+    def get_savetxt_fmt_dict(self):
```
Comment (Rodrigo-Tenorio) [inline pyfstat/mcmc_based_searches.py:3403]: Sweet method duplication :wink:

**Review by Rodrigo-Tenorio — COMMENTED** [review]
I agree with your last changes related to the output. #191

File: pyfstat/grid_based_searches.py [hunk 24]
```diff
@@ -101,14 +101,13 @@ def __init__(
...
-        self.keys = self.input_keys.copy()
+        self.output_keys = self.search_keys.copy()
         if self.BSGL:
             self.detstat = "BSGL"
-            self.keys += ["logBSGL"]
+            self.output_keys += ["logBSGL"]
         else:
             self.detstat = "twoF"
-            self.keys += ["twoF"]
+            self.output_keys += [self.detstat]
```
Comment (Rodrigo-Tenorio) [inline pyfstat/grid_based_searches.py:110]: I love the "inconsistency" of "BSGL -> logBSGL" but "twoF" is just "twoF" :wink:

File: pyfstat/grid_based_searches.py (hunk 6 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/grid_based_searches.py:121]: Ok. Question: Does it make sense to choose automatically from `ComputeFstat` or `SemiCoherentSearch` depending on the number of segments? I guess the answer is "Obviously yes, why would you specify the number of segments otherwise?" and "Yes, there's a logging line declaring which class is being instantiated"

File: pyfstat/grid_based_searches.py [hunk 25]
```diff
@@ -195,7 +194,13 @@ def get_input_data_array(self):
...
             for vals in itertools.product(*coord_arrays):
                 input_data.append(vals)
-            self.input_data = np.array(input_data)
+            input_dtype = np.dtype(
+                {
+                    "names": self.search_keys,
+                    "formats": np.repeat(float, len(self.search_keys)),
+                }
+            )
+            self.input_data = np.array(input_data, dtype=input_dtype)
```
Comment (Rodrigo-Tenorio) [inline pyfstat/grid_based_searches.py:203]: Spooky, what's going on here?

File: pyfstat/grid_based_searches.py (hunk 10 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/grid_based_searches.py:331]: Not sure of what's going on here either. Why do you use `np.dtype`?

File: pyfstat/grid_based_searches.py [hunk 26]
```diff
@@ -359,12 +366,12 @@ def _get_tolerance_from_savetxt_fmt(self):
...
 
-    def save_array_to_disk(self, data):
+    def save_array_to_disk(self):
         logging.info("Saving data to {}".format(self.out_file))
         header = "\n".join(self.output_file_header)
-        header += "\n" + " ".join(self.keys)
-        outfmt = self.get_savetxt_fmt()
-        Ncols = np.shape(data)[1]
+        header += "\n" + " ".join(self.output_keys)
+        outfmt = list(self.get_savetxt_fmt().values())
```
Comment (Rodrigo-Tenorio) [inline pyfstat/grid_based_searches.py:373]: There was an "actually not Ok" regarding a similar line. I'm noticing quite a volume of duplicated lines (that's what we have, I guess...)

File: pyfstat/grid_based_searches.py [hunk 27]
```diff
@@ -750,22 +763,22 @@ def run(self, return_data=False):
...
-                "Will save per-Doppler Fstatmap"
-                " results to {}*.dat".format(self.tCWfilebase)
-            )
+        output_dtype = np.dtype(
+            {
+                "names": self.output_keys,
+                "formats": np.repeat(float, len(self.output_keys)),
+            }
+        )
+        data = np.zeros(len(self.input_data), dtype=output_dtype)
```
Comment (Rodrigo-Tenorio) [inline pyfstat/grid_based_searches.py:774]: Same as before, what?

**Review by Rodrigo-Tenorio — COMMENTED** [review]
Summary: Why the `np.dtype`?

**Review by Rodrigo-Tenorio — COMMENTED** [review]
Regarding `tests`, essentially accommodating the changes and adding some testing functionality for the transient module.

File: pyfstat/grid_based_searches.py (hunk 24 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:110]: Actually this makes sense: in the same notation as the BSGL family, F is essentially logBSG (log Bayes factor for Signal vs pure Gaussian noise). (Never mind the "two".)

File: pyfstat/grid_based_searches.py (hunk 10 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:331]: I think mostly to work nicely with `np.savetxt` and `np.genfromtxt`. Without setting the `dtype` manually, an `np.array` with named columns would always have them all as floats.

File: pyfstat/grid_based_searches.py (hunk 27 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:774]: Same logic as before, but on rereading this, I'm wondering why in this case they're still all `float` and no `int` for the `["t0", "tau"]` parameters... Hmm...

File: pyfstat/grid_based_searches.py (hunk 26 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:373]: for reference, that was this thread: https://github.com/PyFstat/PyFstat/pull/171#discussion_r516936239

File: pyfstat/grid_based_searches.py (hunk 27 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/grid_based_searches.py:774]: I'm not familiar with these parameters.

We use `int` for time whenever time is set using some variation of `LALEpoch` or `LALGPS` or that sort of thing (otherwise parsing becomes a nightmare, as seen multiple times in this package). 

Are these parameters treated as epochs in LALSuite?

Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:774]: ...[no](https://lscsoft.docs.ligo.org/lalsuite/lalpulsar/structtransient_window_range__t.html)...

Comment (Rodrigo-Tenorio) [reply pyfstat/grid_based_searches.py:774]: I see, they are literally what in python would be `int` (if my understanding about swig magic is not mistaken).

Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:774]: ...while [CFSv2](https://git.ligo.org/lscsoft/lalsuite/-/blob/master/lalapps/src/pulsar/Fstatistic/ComputeFstatistic_v2.c#L1086) takes either `transient_t0Epoch`, which is an `EPOCH` type, or `transient_t0Offset`, which is not, and that makes sense. `tau` (duration) by definition is always a length, so doesn't need to be an epoch.

Comment (Rodrigo-Tenorio) [reply pyfstat/grid_based_searches.py:774]:
```suggestion
        output_dtype = np.dtype(
            {
                "names": self.output_keys,
                "formats": np.array([float if key not in ["t0", "tau"] else int for key in self.output_keys]),
            }
        )
        data = np.zeros(len(self.input_data), dtype=output_dtype)
```

File: pyfstat/grid_based_searches.py (hunk 26 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py]: I think it didn't cause any errors for the grid-based searches, but to be on the safe side I've replicated the same fix as for MCMC here too: d6a50ffb771107809400bb71ac6826f402c96851

File: pyfstat/grid_based_searches.py (hunk 27 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:783]: This does work, but has one problem: if an existing data file is found, `np.genfromtxt` (called via `helper_functions.read_txt_file_with_header`) guesses the dtypes as float (even if in the text file none of the t0/tau colum entries have any decimals, as our _output_ format for those is already `%d`). So in that case, `self.data = old_data` will have all-float dtype, while if an actual search is run, it will have `[float,float,...,int,int]`. If one just runs a search and writes the output to disk, there's no problem, as the mentioned output formatting will make the difference disappear. But if someone works with the search class interactively, this might cause some confusion.

File: pyfstat/grid_based_searches.py (hunk 10 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:331]: Testing and researching a bit more: and the specific reason for why I explicitly set the `formats` (i.e. actual data type) of the `dtype` as well is that I can't give it `names` alone, nor define a plain `np.array` and later change its `data.dtype.names` (-> "ValueError: there are no fields defined").

Alternatives would be to use `pandas`, or simple dicts of arrays (which would be more analogous to the MCMC class, so maybe actually a good idea!).

File: pyfstat/grid_based_searches.py (hunk 27 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:783]: We decided for now to leave it as is: no real problems from having them all stored as float internally. The whole thing can always get improved further in the future, but then we may want to switch to a unified dict-based storage for both grid- and MCMC-searches anyway, making extra work on this superfluous for now.

File: pyfstat/grid_based_searches.py (hunk 25 above)
Comment (dbkeitel) [reply pyfstat/grid_based_searches.py:203]: see the other threads

**Review by Rodrigo-Tenorio — APPROVED** [review]

**Comment by dbkeitel** [comment]
examples look fine: https://github.com/PyFstat/PyFstat/actions/runs/452416869
